cloud-aws/src/main/java/com/sequenceiq/cloudbreak/cloud/aws/AwsMetadataCollector.java

```diff
@@ -60,53 +65,64 @@
             AmazonEC2Client amazonEC2Client =
                     awsClient.createAccess(new AwsCredentialView(ac.getCloudCredential()), region);
 
-            //contains all instances
-            ListMultimap<String, CloudInstance> groupByInstanceGroup = groupByInstanceGroup(vms);
+            List<String> knownInstanceIdList = allInstances.stream()
+                    .filter(cloudInstance -> cloudInstance.getInstanceId() != null)
+                    .map(CloudInstance::getInstanceId)
+                    .collect(Collectors.toList());
 
-            for (String key : groupByInstanceGroup.keySet()) {
-                List<CloudInstance> cloudInstances = groupByInstanceGroup.get(key);
-                cloudVmMetaDataStatuses.addAll(collectGroupMetaData(ac, amazonASClient, amazonEC2Client, amazonCFClient, key, cloudInstances));
+            Multimap<InstanceTemplate, CloudInstance> instanceTemplateList = ArrayListMultimap.create();
+            Map<InstanceTemplate, InstanceAuthentication> templateInstanceAuthenticationMap = new HashMap<>();
+            for (CloudInstance vm : vms) {
+                templateInstanceAuthenticationMap.put(vm.getTemplate(), vm.getAuthentication());
+                instanceTemplateList.put(vm.getTemplate(), vm);
             }
 
-            return cloudVmMetaDataStatuses;
+            for (InstanceTemplate instanceTemplate : instanceTemplateList.keySet()) {
```

Comment:
The evaluation of unknown instances could be in an individual method.

Reply:
fixed